sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/BatchGroupUtils.scala

```diff
@@ -392,40 +393,80 @@ private[python] object BatchGroupedIterator {
  * @param numOutputBatches a metric for output batches
  */
 class CombiningIterator(
-    inputBatchQueue: BatchQueue,
+    inputBatchQueue: IteratorBatchQueue,
     pythonOutputIter: Iterator[ColumnarBatch],
     pythonArrowReader: GpuPythonArrowOutput,
     numOutputRows: GpuMetric,
     numOutputBatches: GpuMetric) extends Iterator[ColumnarBatch] {
 
-  // For `hasNext` we are waiting on the queue to have something inserted into it
-  // instead of waiting for a result to be ready from Python. The reason for this
-  // is to let us know the target number of rows in the batch that we want when reading.
-  // It is a bit hacked up but it works. In the future when we support spilling we should
-  // store the number of rows separate from the batch. That way we can get the target batch
-  // size out without needing to grab the GpuSemaphore which we cannot do if we might block
-  // on a read operation.
-  override def hasNext: Boolean = inputBatchQueue.hasNext || pythonOutputIter.hasNext
+  private var pending: Option[SpillableColumnarBatch] = None
+  Option(TaskContext.get()).foreach(onTaskCompletion(_)(pending.foreach(_.close())))
+
+  override def hasNext: Boolean = pythonOutputIter.hasNext
```

Comment:
This is rather confusing, seeing how there is a `pending` but it is not included in the `hasNext`. It would be good to rename pending to make it clear that it is `pendingInput` and have a comment explaining how it is only for the input side and the python output should line up row for row so we only look at the python output iterator.

Reply:
Done